rfcs/shared/build-system/03-packaging-for-npm.md

```diff
@@ -172,25 +172,88 @@ This approach significantly improves our current situation:
   - non explicit for human (if `files` property is used, some files will be automatically and always shipped)
   - 2 sources of "truth" (`.npmignore` and `files`)
 
-### 2. shipping only type definition rollup
+### 2. shipping type declarations (TypeScript)
 
-> **NOTE:** to make this RFC more focused this will be covered in followup PR's if needed
 
 #### Problem statement
 
-- TBA
+To be able to provide TypeScript support to our consumers we generate declaration files during `tsc` invocation.
+
+These declaration files are:
+
+1. generated side by side with implementation files
+1. generated N times for N supported module loaders (for v9 N equals to 2 - esm and commonjs)
+
+Aforementioned approach introduces 2 problems.
+
+**Problem 1: generated side by side with implementation files**
+
+- exposure of all package internals and structure and allows consumers to use **private APIs**
+- exposure of all TypeScript constructs that might violate consumers TypeScript version restrictions
+  - eg. internally in the package we use Template literal types that are not exposed to public API. If users depend on TS version that doesn't support this language feature they will get TS errors thus won't be able to ship code.
+
+**Problem 2: generated N times for N supported module loaders**
+
+- generated TypeScript declaration files don't emit different output based on different `module` field, thus this generation is redundant and provides no benefit.
+
+_Example:_
+
+_`module: CommonJs`_
+
+```ts
+import type { FunctionComponent } from 'react';
+import { TextWrapperProps } from '../wrapper';
+export declare const bodyClassName = 'fui-Body';
+export declare const Body: FunctionComponent<TextWrapperProps>;
+```
+
+_`module: ESM`_
+
+```ts
+import type { FunctionComponent } from 'react';
+import { TextWrapperProps } from '../wrapper';
+export declare const bodyClassName = 'fui-Body';
+export declare const Body: FunctionComponent<TextWrapperProps>;
+```
 
 #### Detailed Design or Proposal
 
-- TBA
+**Problem 1 solution:**
+
+- ship only [rolluped declaration files `.d.ts`](https://medium.com/@martin_hotell/typescript-library-tips-rollup-your-types-995153cc81c7) for all public API entrances. We can leverage already existing setup via api-extractor.
```

Comment:
The diagram mentions rollup, but the solution using api-extractor is still comptible with how we currently build with tsc right ?